misc/testerina/modules/testerina-core/src/main/ballerina/execute.bal

```diff
@@ -36,56 +40,107 @@ public function startSuite() returns int {
     } else {
         if testRegistry.getFunctions().length() == 0 && testRegistry.getDependentFunctions().length() == 0 {
             println("\tNo tests found");
-            return exitCode;
+            lock {
+                return exitCode;
+            }
+
         }
 
         error? err = orderTests();
         if err is error {
-            exitCode = 1;
+            enableExit();
             println(err.message());
         } else {
             executeBeforeSuiteFunctions();
-            executeTests();
+            err = executeTests();
+            if err is error {
+                enableExit();
+                println(err.message());
+            }
             executeAfterSuiteFunctions();
             reportGenerators.forEach(reportGen => reportGen(reportData));
         }
     }
-    return exitCode;
+    lock {
+        return exitCode;
+    }
 }
 
-function executeTests() {
+function executeTests() returns error? {
+    decimal startTime = currentTimeInMillis();
     foreach TestFunction testFunction in testRegistry.getFunctions() {
-        executeTest(testFunction);
+        _ = testFunction.parallelizable ? conMgr.addInitialParallelTest(testFunction) : conMgr.addInitialSerialTest(testFunction);
+    }
+
+    while !conMgr.isExecutionDone() {
+
+        if conMgr.getAvailableWorkers() != 0 {
+            conMgr.waitUntilEmptyQueueFilled();
+
+            if conMgr.getSerialQueueLength() != 0 && conMgr.getAvailableWorkers() == testWorkers {
+                TestFunction testFunction = conMgr.getSerialTest();
+                conMgr.addTestInExecution(testFunction);
+                conMgr.allocateWorker();
+                future<error?> serialWaiter = start executeTest(testFunction);
+                any _ = check wait serialWaiter;
+
+            }
+
+            else if conMgr.getParallelQueueLength() != 0 && conMgr.getSerialQueueLength() == 0 {
+                TestFunction testFunction = conMgr.getParallelTest();
+                conMgr.addTestInExecution(testFunction);
+                conMgr.allocateWorker();
+                future<(error?)> parallelWaiter = start executeTest(testFunction);
+                if isDataDrivenTest(testFunction) {
+                    any _ = check wait parallelWaiter;
```

Comment:
Why do you need this? May help to add some comments in some places too.

Reply:
For data driven tests, Each data entry of the test can be executed in parallel before starting next test function. Here, we are waiting for the complete execution of the data driven test.